src/main/scala/com/snowplowanalytics/refererparser/scala/Parser.scala

```diff
@@ -1,77 +1,117 @@
+  * You may obtain a copy of the License at
+  *
+  *      http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
 package com.snowplowanalytics.refererparser.scala
 
 // Java
+import java.io.ByteArrayInputStream
 import java.net.{URI, URISyntaxException}
 
+import scala.util.Failure
+
 // RefererParser Java impl
 import com.snowplowanalytics.refererparser.{Parser => JParser}
 import com.snowplowanalytics.refererparser.{Medium => JMedium}
 
 // Scala
-import scala.collection.JavaConversions._ 
+import scala.collection.JavaConversions._
+import scala.util.Try
+import com.typesafe.scalalogging.Logger
 
 /**
- * Enumeration for supported mediums.
- *
- * Replacement for Java version's Enum.
- */
+  * Enumeration for supported mediums.
+  *
+  * Replacement for Java version's Enum.
+  */
 object Medium extends Enumeration {
   type Medium = Value
 
-  val Unknown  = Value("unknown")
-  val Search   = Value("search")
+  val Unknown = Value("unknown")
+  val Search = Value("search")
   val Internal = Value("internal")
-  val Social   = Value("social")
-  val Email    = Value("email")
+  val Social = Value("social")
+  val Email = Value("email")
+  val Paid = Value("paid")
 
   /**
-   * Converts from our Java Medium Enum
-   * to our Scala Enumeration values above.
-   */
+    * Converts from our Java Medium Enum
+    * to our Scala Enumeration values above.
+    */
   def fromJava(medium: JMedium) =
     Medium.withName(medium.toString())
 }
 
 /**
- * Immutable case class to hold a referer.
- *
- * Replacement for Java version's POJO.
- */
+  * Immutable case class to hold a referer.
+  *
+  * Replacement for Java version's POJO.
+  */
 case class Referer(
-  medium: Medium.Medium,
-  source: Option[String],
-  term:   Option[String]
+    medium: Medium.Medium,
+    source: Option[String],
+    term: Option[String]
 )
 
 /**
- * Parser object - contains one-time initialization
- * of the YAML database of referers, and parse()
- * methods to generate a Referer object from a
- * referer URL.
- *
- * In Java this had to be instantiated as a class.
- */
+  * Parser object - contains one-time initialization
+  * of the YAML database of referers, and parse()
+  * methods to generate a Referer object from a
+  * referer URL.
+  *
+  * In Java this had to be instantiated as a class.
+  */
 object Parser {
 
   type MaybeReferer = Option[Referer]
 
-  private lazy val jp = new JParser()
+  val logger = Logger[Parser.type]
+
+  private lazy val jp: JParser = createParser
+
+  val createParser: JParser = {
+    import org.apache.http.client.fluent.Request
+    import org.apache.http.util.EntityUtils
```

Comment:
Why are these 2 imports in here and not at the top with the others? For example `Logger` is at the top so just wondering how you decide where you put an import